File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys
input = sys.stdin.readline

k, n = map(int, input().rstrip().split())
cable = [int(input()) for _ in range(k)]
short, long = 1, max(cable)

while short <= long:
    cnt = 0
    mid = (short + long) // 2

    for lan in cable:
        cnt += lan // mid

    if cnt < n:
        long = mid - 1
    else:
        short = mid + 1

print(long)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.util.*;
import java.io.*;

public class Main {
    static int K, N;
    static Long[] Cables;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        K = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());
        Cables = new Long[K];

        for (int i = 0; i < K; i++) {
            Cables[i] = Long.parseLong(br.readLine());
        }

        Arrays.sort(Cables);

        long max = Cables[K - 1];
        long min = 1;

        while(min <= max) {
            long cnt = 0;
            long mid = (max + min) / 2;

            for (long i : Cables) {
                cnt += i / mid;
            }

            if (cnt < N) {
                max = mid - 1;
            } else {
                min = mid + 1;
            }
        }

        System.out.println(min - 1);
    }
}